src/cockpit/channels/metrics.py

```diff
@@ -15,23 +15,192 @@
+        interval = options.get('interval', self.interval)
+        if interval <= 0 or interval > sys.maxsize:
+            logger.error('invalid "interval" value: %s', interval)
+            self.close(problem="protocol-error", message='invalid interval option')
+            return False
+
+        self.interval = interval
+
+        metrics = options.get('metrics')
+        if not metrics or not isinstance(metrics, list):
+            logger.error('invalid "metrics" value: %s', metrics)
+            self.close(problem='metrics is not an list')
+            return False
+
+        for metric in metrics:
+            # validate it's an object
+            name = metric.get('name')
+            units = metric.get('units')
+            derive = metric.get('derive')
+
+            desc = SAMPLE_DESCRIPTIONS.get(name)
+            if not desc:
+                logger.error('unsupported metric: %s', name)
+                self.close(problem='not-supported',
+                           message='invalid metric')
+                return False
+
+            (unit, semantics, instanced, sampler) = desc
+            if units and units != unit:
+                self.close(problem='protocol-error',
+                           message=f'{name} has units {unit}, not {units}')
+                return False
+
+            # Keep track of what samplers are requested
+            self.samplers |= sampler.value
+
+            desc = MetricDescription(name, unit, semantics, instanced, sampler.value)
+            self.metrics.append(MetricInfo(desc, derive, {}, None, None))
+
+        return True
+
+    def cockpit_internal_metrics_sample(self, metric, instance=None, value=None):
+        for index, metricinfo in enumerate(self.metrics):
+            if metricinfo.desc.name != metric:
+                continue
+
+            if metricinfo.desc.instanced:
+                inst = metricinfo.instances.get(instance)
+                if inst is None:
+                    self.need_meta = True
+
+                inst = Instance(value=value,
+                                old_value=inst.value if inst else None,
+                                seen=True)
+                instances = metricinfo.instances
+                instances[instance] = inst
+                self.metrics[index] = MetricInfo(desc=metricinfo.desc, derive=metricinfo.derive, instances=instances,
+                                                 value=value, old_value=metricinfo.value)
+            else:
+                self.metrics[index] = MetricInfo(desc=metricinfo.desc, derive=metricinfo.derive, instances={},
+                                                 value=value, old_value=metricinfo.value)
+
+    def send_updates(self):
+        data = []
+        metrics = []
+        timestamp = int(time.time())
+        self.next_timestamp = timestamp
+
+        if self.samplers & Sampler.CPU_SAMPLER.value:
+            cockpit_cpu_samples(self.cockpit_internal_metrics_sample)
```

Comment:
I'm not a fan of passing in a bound callback to the samples function.  Why not either:

 - add the sampler functions directly on the metrics class itself; or
 - define some sort of an interface for setting samples data and pass an object which implements that interface (possibly the samples object itself)

again: the naming here is a bit too "C" and not enough "py"

Reply:
1. I would love to keep the file small, honestly rename it to `internalmetrics.py`, but hmm seems doable, it does mean the class will be filled with:
* cpu
* memory
* cgroups
* disk

Reply:
Ok, we just keep two dicts: one for the old values and one for new.